### PR TITLE
search: handle hidden records correctly

### PR DESCRIPTION
**Pull request by MJedr**
ref: #1739

### CODE REVIEW

File: backend/inspirehep/access_control.py [hunk 1]
```diff
@@ -37,6 +37,17 @@ def can(self):
 
 class LiteraturePermissionCheck(SessionCatalogerPermission):
     def can(self):
-        if "Literature" not in self.record.get("_collections", []):
+        collections = {
```
Comment (michamos) [inline backend/inspirehep/access_control.py:40]: maybe make it clear what these collections are?
```suggestion
        non_private_collections = {
```

File: backend/inspirehep/search/api.py [hunk 2]
```diff
@@ -186,10 +186,9 @@ def execute(self, *args, **kwargs):
         return results
 
     def query_for_superuser_or_users(self, query_string):
+        collection = request.args.get("collection", default="literature", type=str)
         if not is_superuser_or_cataloger_logged_in():
-            user_query = Q(
-                IQ(query_string, self) & Q("term", _collections="Literature")
-            )
+            user_query = Q(IQ(query_string, self) & Q("term", _collections=collection))
```
Comment (michamos) [inline backend/inspirehep/search/api.py:191]: what prevents a non-priviledged user from searching in private collections now, by setting something like `collection=d0 internal notes`?

File: backend/inspirehep/search/api.py [hunk 3]
```diff
@@ -187,9 +192,29 @@ def execute(self, *args, **kwargs):
 
     def query_for_superuser_or_users(self, query_string):
         if not is_superuser_or_cataloger_logged_in():
-            user_query = Q(
-                IQ(query_string, self) & Q("term", _collections="Literature")
+            non_private_collections = [
```
Comment (michamos) [inline backend/inspirehep/search/api.py:195]: This is used in two places now. Maybe move it to the config?

File: backend/inspirehep/search/api.py [hunk 4]
```diff
@@ -187,9 +192,29 @@ def execute(self, *args, **kwargs):
...
+                "Fermilab",
+                "HAL Hidden",
+                "LArSoft Notes",
+                "SLAC",
+                "ZEUS Preliminary Notes",
+            ]
+            validation_regex = f"_collections: *({'|'.join(non_private_collections)})"
+            matched_collections = re.match(
+                validation_regex, query_string.lower(), re.IGNORECASE
             )
```
Comment (michamos) [inline backend/inspirehep/search/api.py:209]: 1. I'm confused, I thought the idea was to use the `collection` request arg? why are you trying to get the collection from the query string here? I thought the idea was that by default you search in `Literature`, and if a `collection` is passed you search in that collection?
2. Also this is a rather bad way of restricting to allowed collections, as you're relying on one specific spelling of the query, which doesn't necessarily match what the query parser understands. So the correct way to do this is to try to extract the collections from the query parser (which is the thing that knows how to parser a query), but I don't think you actually need to do that due to 1., as it's easy to figure out if a collection is allowed if you don't need to extract it from the query string.

Comment (MJedr) [reply backend/inspirehep/search/api.py:209]: No, actually with @drjova we discussed that the first idea is too hackish (from the frontend side) and will require many changes and we decided to do it in a way it's handled on legacy.

File: backend/tests/integration/search/test_api.py [hunk 5]
```diff
@@ -1052,3 +1052,69 @@ def test_public_api_generates_correct_links_in_literature_search_with_fields(
...
+            "_collections": ["D0 Preliminary Notes"],
+            "titles": [{"title": "A hidden record"}],
+        },
+    )
+    with inspire_app.test_client() as client:
+        url = "/api/literature?q=_collections%3A%20fermilab%20or%20_collections%3A%20HAL%20hidden%20or%20_collections%3A%20Literature"
+        response = client.get(url)
+
+    assert response.status_code == 200
+    assert len(response.json["hits"]["hits"]) == 3
```
Comment (michamos) [inline backend/tests/integration/search/test_api.py:1120]: can you also add a test that makes sure we never return records from private collactions?

File: backend/tests/integration/records/views/test_views_literature.py [hunk 6]
```diff
@@ -850,3 +850,16 @@ def test_literature_json_put_redirected_record(inspire_app):
...
+    record = create_record(
+        "lit",
+        data={
+            "_collections": ["Fermilab"],
+            "titles": [{"title": "A literature record from fermilab collection"}],
+        },
+    )
+    with inspire_app.test_client() as client:
+        response = client.get(f"/literature/{record.control_number}")
+    assert response.status_code == 200
```
Comment (michamos) [inline backend/tests/integration/records/views/test_views_literature.py:865]: do we have a test that makes sure users can't access private records?

**Review by michamos — APPROVED** [review]

File: backend/inspirehep/search/api.py [hunk 7]
```diff
@@ -187,10 +187,19 @@ def execute(self, *args, **kwargs):
...
+                    "terms",
+                    _collections=current_app.config[
+                        "NON_PRIVATE_LITERATURE_COLLECTIONS"
+                    ],
+                )
+            else:
+                user_query = Q(
+                    IQ(query_string, self) & Q("term", _collections="Literature")
+                )
+                return self.query(user_query)
```
Comment (drjova) [inline backend/inspirehep/search/api.py:202]:
```suggestion
            if "_collections" in query_string:
                user_query = Q(IQ(query_string, self))
                return self.query(user_query).filter(
                    "terms",
                    _collections=current_app.config[
                        "NON_PRIVATE_LITERATURE_COLLECTIONS"
                    ],
                )
            
            user_query = Q(
                IQ(query_string, self) & Q("term", _collections="Literature")
            )
            return self.query(user_query)
```

**Review by drjova — APPROVED** [review]
💯  just the comment :)

**Pull request merged**